src/source/context.rs

```diff
@@ -22,3 +25,44 @@ impl Default for GlobalContext {
         Self
     }
 }
+
+pub struct CreatableSourceContext<'a, D> {
+    source: *mut obs_source_t,
+    pub(crate) hotkey_callbacks: Vec<(
+        ObsString,
+        ObsString,
+        Box<dyn FnMut(&mut Hotkey, &mut Option<D>)>,
+    )>,
+    pub settings: DataObj<'a>,
+    pub global: &'a mut GlobalContext,
+}
+
+impl<'a, D> CreatableSourceContext<'a, D> {
+    pub(crate) unsafe fn from_raw(
+        source: *mut obs_source_t,
+        settings: DataObj<'a>,
+        global: &'a mut GlobalContext,
+    ) -> Self {
+        Self {
+            source,
+            hotkey_callbacks: Default::default(),
+            settings,
+            global,
+        }
+    }
+
+    pub fn register_hotkey(
+        &mut self,
+        name: ObsString,
+        description: ObsString,
+        func: Box<dyn FnMut(&mut Hotkey, &mut Option<D>)>,
+    ) {
+        self.hotkey_callbacks.push((name, description, func));
+    }
```

Comment:
```suggestion
    pub fn register_hotkey<F: FnMut(&mut Hotkey, &mut Option<D>) + 'static>(
        &mut self,
        name: ObsString,
        description: ObsString,
        func: F,
    ) {
        self.hotkey_callbacks
            .push((name, description, Box::new(func)));
    }
```
Requiring the user to box them looks bad.